src/common/storage/backends/s3/Client.js

```diff
@@ -0,0 +1,178 @@
+        this.s3Client = null;
+        this.ready = this.initialize();
+    };
+
+    S3Storage.prototype = Object.create(StorageClient.prototype);
+
+
+    S3Storage.prototype.initialize = async function () {
+        if (require.isBrowser) {
+            await new Promise((resolve, reject) => {
+                require(['aws-sdk-min'], () => {
+                    this.AWS = window.AWS;
+                    resolve();
+                }, reject);
+            });
+        } else {
+            this.AWS = require.nodeRequire('aws-sdk');
+        }
+    };
+
+    S3Storage.prototype.initializeS3Client = function (AWS, config) {
+        const s3Client = new AWS.S3(config);
+        promisifyMethod(s3Client, 'createBucket');
+        promisifyMethod(s3Client, 'getObject');
+        promisifyMethod(s3Client, 'putObject');
+        promisifyMethod(s3Client, 'deleteObject');
+        promisifyMethod(s3Client, 'headObject');
+        promisifyMethod(s3Client, 'listObjectsV2');
+        return s3Client;
+    };
+
+    S3Storage.prototype.getS3Client = async function (config) {
+        await this.ready;
+        if(!config){
+            if(!this.s3Client){
+                this.s3Client = this.initializeS3Client(this.AWS, this.config);
+            }
+            return this.s3Client;
+        } else {
+            config = this.createS3Config(config);
+            return this.initializeS3Client(this.AWS, config);
+        }
+    };
+
+    S3Storage.prototype.createS3Config = function (config) {
+        return {
+            endpoint: config.endpoint || 'http://localhost:80',
+            accessKeyId: config.accessKeyId,
+            secretAccessKey: config.secretAccessKey,
+            sslEnabled: config.endpoint ? config.endpoint.startsWith('https') : false,
+            s3ForcePathStyle: true, // needed with minio, from the documentation
+            signatureVersion: 'v4'
+        };
+    };
+
+    S3Storage.prototype.createBucketIfNeeded = async function () {
+        const s3Client = await this.getS3Client();
+        try {
+            await s3Client.createBucket({
+                Bucket: this.bucketName
+            });
+        } catch (err) {
+            if (err['statusCode'] !== 409) {
+                this.logger.error(`Failed to create bucket ${this.bucketName} in S3 server.`);
+                throw err;
+            }
+
+        }
+    };
+
+    S3Storage.prototype.getFile = async function (dataInfo) {
+        const {endpoint, bucketName, filename} = dataInfo.data;
+        const {accessKeyId, secretAccessKey} = this.config;
+        const s3Client = await this.getS3Client({endpoint, accessKeyId, secretAccessKey});
+        const data = await s3Client.getObject({
+            Bucket: bucketName,
+            Key: filename
+        });
+        return data.Body;
+    };
+
+    S3Storage.prototype.putFile = async function (filename, content) {
+        await this.createBucketIfNeeded();
```

Comment:
As a very minor point, it might be nice to pass the s3Client to this argument to make it clear that it is using the same one as below. Not necessary though - just a comment for clarity.

Reply:
Yeah, this is a good point. This will make more sense if this code is ever refactored such that `createBucketIfNeeded` is being used by Multiple other methods. I will refactor `createBucketIfNeeded` to take in an s3Client object.

Reply:
`S3Storage.prototype.createBucketIfNeeded` now takes a `s3Client` object.